### PR TITLE
Correct getAccessStrategy for class extension storage decls.

### PR DESCRIPTION
**Pull request by jckarter**
Class extension methods are non-polymorphic unless @objc. Fixes rdar://problem/32434652.

### CODE REVIEW

**Comment by jckarter** [comment]
@slavapestov @jrose-apple Are there any other edge cases to account for here?

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
LGTM.

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -1165,7 +1165,13 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
 
   case DeclKind::Class:
     // Final properties can always be direct, even in classes.
-    return !storage->isFinal();
+    if (storage->isFinal())
+      return false;
+    // Extension properties are statically dispatched, unless they're @objc.
+    if (storage->getDeclContext()->isExtensionContext()
+        && !storage->isObjC())
```
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: This seems like a coincidence rather than a design decision. I'd rather not put that into key path logic.

Comment (jrose-apple) [inline lib/AST/Decl.cpp]: I fully expect us to get overridable methods in extensions some day, and that will include computed properties.

Comment (jrose-apple) [inline lib/AST/Decl.cpp]: Also, is there a distinction to be made between "imported from ObjC" and "declared `@objc`" and "`@objc dynamic`"?

Comment (jckarter) [inline lib/AST/Decl.cpp]: This isn't keypath-specific, it's part of `AbstractStorageDecl::getAccessStrategy`. When we change the language, this seems like the right place to centralize that logic.

Comment (jrose-apple) [inline lib/AST/Decl.cpp]: I'd rather just be implicitly marking extension declarations `final` than put special cases for them all over the place. This might be the *right* place to put one, but it's not the only one.

Comment (jckarter) [inline lib/AST/Decl.cpp]: I agree that making them `final` is a better fix, but might be a bit too invasive for 4.0. Do we have a bug for that?
